src/network/network_crypto.h

```diff
@@ -0,0 +1,183 @@
+ */
+class NetworkEncryptionHandler {
+public:
+	virtual ~NetworkEncryptionHandler() {}
+
+	/**
+	 * Get the size of the MAC (Message Authentication Code) used by the underlying encryption protocol.
+	 * @return The size, in bytes, of the MACs.
+	 */
+	virtual size_t MACSize() const = 0;
+
+	/**
+	 * Decrypt the given message in-place, validating against the given MAC.
+	 * @param mac The message authentication code (MAC).
+	 * @param message The location of the message to decrypt.
+	 * @return Whether decryption and authentication/validation of the message succeeded.
+	 */
+	virtual bool Decrypt(std::span<std::uint8_t> mac, std::span<std::uint8_t> message) = 0;
+
+	/**
+	 * Encrypt the given message in-place, and write the associated MAC.
+	 * @param mac The location to write the message authentication code (MAC) to.
+	 * @param message The location of the message to encrypt.
+	 */
+	virtual void Encrypt(std::span<std::uint8_t> mac, std::span<std::uint8_t> message) = 0;
+};
+
+/** The authentication method that can be used. */
+enum NetworkAuthenticationMethod : uint8_t {
+	NETWORK_AUTH_METHOD_X25519_PLAIN, ///< No actual authentication is taking place, just perform a x25519 key exchange.
+	NETWORK_AUTH_METHOD_X25519_PAKE, ///< Authentication using x25519 password-authenticated key agreement.
+	NETWORK_AUTH_METHOD_END, ///< Must ALWAYS be on the end of this list!! (period)
+};
+
+/** The mask of authentication methods that can be used. */
+using NetworkAuthenticationMethodMask = uint16_t;
+
+/**
+ * Base class for cryptographic authentication handlers.
+ */
+class NetworkAuthenticationHandler {
+public:
+	virtual ~NetworkAuthenticationHandler() {}
+
+	/**
+	 * Get the name of the handler for debug messages.
+	 * @return The name of the handler.
+	 */
+	virtual std::string_view GetName() = 0;
+
+	/**
+	 * Get the method this handler is providing functionality for.
+	 * @return The \c NetworkAuthenticationMethod.
+	 */
+	virtual NetworkAuthenticationMethod GetAuthenticationMethod() = 0;
+
+	/**
+	 * Create a \a NetworkEncryptionHandler to encrypt or decrypt messages from the client to the server.
+	 * @return The handler for the client to server encryption.
+	 */
+	virtual std::unique_ptr<NetworkEncryptionHandler> CreateClientToServerEncryptionHandler() const = 0;
+
+	/**
+	 * Create a \a NetworkEncryptionHandler to encrypt or decrypt messages from the server to the client.
+	 * @return The handler for the server to client encryption.
+	 */
+	virtual std::unique_ptr<NetworkEncryptionHandler> CreateServerToClientEncryptionHandler() const = 0;
+};
+
+/**
+ * Base class for client side cryptographic authentication handlers.
+ */
+class NetworkAuthenticationClientHandler : public NetworkAuthenticationHandler {
+public:
+	/** The processing result of receiving a request. */
+	enum RequestResult {
+		AWAIT_USER_INPUT, ///< We have requested some user input, but must wait on that.
+		IMMEDIATE_RESPONSE, ///< We can immediately respond to the server.
```

Comment:
This name took me a bit to understand, but I guess it is meant as counter balance to `AWAIT`. Maybe `READY`? Not sure if that is any better :)

Reply:
It is indeed to have something different that `AWAIT`, but just like you I having problems giving it a sensible name. `READY` implies that all work is already done, while it is to instruct that the network code should just send the response instead of doing it via the callback.

"There are 2 hard problems in computer science: cache invalidation, naming things, and off-by-1 errors."

Reply:
`READY_FOR_RESPONSE`?

I think the "immediate" is mostly what is a bit unclear to me. Like: ANSWER NOW FAST FAST, I got that impression. But it is more meant as: either we are waiting for more user input, or we are ready to go to the next step, right?

And yes, naming is hard :)